src/Primitives/arduino.cpp

```diff
@@ -1026,3 +1074,44 @@ bool resolve_external_memory(char *symbol, Memory **val) {
     FATAL("Could not find memory %s \n", symbol);
     return false;
 }
+
+//------------------------------------------------------
+// Restore external state when restoring a snapshot
+//------------------------------------------------------
+void restore_external_state(Module *m,
+                            std::vector<IOStateElement> external_state) {
+    uint8_t opcode = *m->pc_ptr;
+    // TODO: Maybe primitives can also be called using the other call
+    // instructions such as call_indirect
+    //  maybe there should just be a function that checks if a certain function
+    //  is being called that handles all these cases?
+    if (opcode == 0x10) {  // call opcode
```

Comment:
I am surprised that we check an opcode here. Why is that?

Reply:
What we do here is check if we are stepping back to a point where we called a function. If we did call a function we check which function that was. If it was a primitive we look up the action needed to reverse the action performed by the primitive.

You can theoretically also do this without these checks and just execute the restore operation on all primitives that have it but that can sometimes be incorrect depending on which primitives you have. For example if you have a primitive to drive one motor x degrees or a primitive to drive two motors x degrees at the same time. The result of restoring the motor encoder positions one motor at a time or both motors at the same time is different. So it makes sense so look up the associated function and execute that one so you actually execute the reverse of what you just did.

Reply:
I don't understand how the restore operation on all primitives could have different behaviour. We should discuss this further offline.